lib/galaxy/datatypes/images.py

```diff
@@ -245,25 +271,23 @@ def set_meta(
                         "num_unique_values",
                     ]
                 }
-                for page in tif.series:
+
+                # TIFF files can contain multiple images, each represented by a series of pages
+                for series in tif.series:
 
                     # Determine the metadata values that should be generally available
-                    metadata["axes"].append(page.axes.upper())
-                    metadata["dtype"].append(str(page.dtype))
+                    metadata["axes"].append(series.axes.upper())
+                    metadata["dtype"].append(str(series.dtype))
 
                     axes = metadata["axes"][-1].replace("S", "C")
-                    metadata["width"].append(Tiff._get_axis_size(page.shape, axes, "X"))
-                    metadata["height"].append(Tiff._get_axis_size(page.shape, axes, "Y"))
-                    metadata["channels"].append(Tiff._get_axis_size(page.shape, axes, "C"))
-                    metadata["depth"].append(Tiff._get_axis_size(page.shape, axes, "Z"))
-                    metadata["frames"].append(Tiff._get_axis_size(page.shape, axes, "T"))
+                    metadata["width"].append(Tiff._get_axis_size(series.shape, axes, "X"))
+                    metadata["height"].append(Tiff._get_axis_size(series.shape, axes, "Y"))
+                    metadata["channels"].append(Tiff._get_axis_size(series.shape, axes, "C"))
+                    metadata["depth"].append(Tiff._get_axis_size(series.shape, axes, "Z"))
+                    metadata["frames"].append(Tiff._get_axis_size(series.shape, axes, "T"))
 
                     # Determine the metadata values that require reading the image data
-                    try:
-                        im_arr = page.asarray()
-                        metadata["num_unique_values"].append(len(np.unique(im_arr)))
-                    except ValueError:  # Occurs if the compression of the TIFF file is unsupported
-                        pass
+                    metadata["num_unique_values"].append(Tiff._get_num_unique_values(series))
```

Comment:
This is still very slow, I have a 600KB tiff file (see https://usegalaxy.org/u/marius/h/slow-tiff) where this takes more than 10 minutes. It's used in a workflow that can complete in under 5 minutes and it's been going for hours now.

Reply:
Thanks for spotting this issue! There was a bug in the implementation. The fix is simple and I'm proposing to merge the fixed version here: https://github.com/galaxyproject/galaxy/pull/20669